compiler/src/dotty/tools/dotc/ast/TreeInfo.scala

```diff
@@ -790,10 +790,13 @@ trait TypedTreeInfo extends TreeInfo[Type] { self: Trees.Instance[Type] =>
 
   /** An extractor for def of a closure contained the block of the closure. */
   object closureDef {
+    private def isDefaultArg(tree:Tree): Boolean = tree match
```

Comment:
I don't think we should mix default arguments with this extractor. I prefer to leave it as it was before.

Reply:
>  just because it is more convenient for some lint. ...

I also cared about it, but when I added `isDefaultArg`, I thought `Block(Nil, expr)` here is the special case for closure where default argument `stmts` is empty.

Do you think the `// closure 1` in the following example is NOT closure?

`Block(Nil, expr)`

```scala
{
  {
      // closure 0
  }
}
```

`Block(stmts, expr)`

```scala
{
  // synthesized ValDefs from default arguments
  val meth$default0 = ???
  val meth$default1 = ???
  // ...
  {
     // closure 1
  } 

}
```

Reply:
I understand, but I don't think we should generalize `closureDef` that way. 

> I also cared about it and when I added isDefaultArg, I thought Block(Nil, expr) here is the special case for closure where default argument stmts is empty, therefore this change would be safe.

That's one of the scenarios. But the compiler is also free to wrap closureDef in a block in other situations.

Reply:
Sure, I will change it back. Thank you for feedback.

What do you think about prohibiting impure lambda in statement position?

https://github.com/lampepfl/dotty/pull/19540#issuecomment-1962119005

Reply:
We could try that and see what breaks. The comments on #11769 indicate that this won't be easy, though.